### ISSUE TITLE
Lost connection to scheduler process on login node (HTCondor)

### ISSUE THREAD

**Issue by simone-codeluppi** [dask/distributed#4654]
**What happened**:
 
Hi 

I am processing 800-1000 3D microscopy images using `dask/dask-jobqueue` on an `htcondor` managed cluster. Each 3D image is processed in parallel (embarrassingly parallel application using mapping functions and gathering futures). Each image is processed in a separated `job` managed by a `worker`. The min and max number of `jobs/workers` is regulated by adaptive scaling (min, max). The expected processing time of a task is quite long (~1hr)

The processing time of each task increases over time and usually the last tasks take hrs (~5-10hrs) to be completed or the processing stalls. The tasks that are run for longer time or stall are random (different images can be processed last). This is similar to what is reported here: [#2835](https://github.com/dask/distributed/issues/2835) but having `ctoolz` didn't solve the issue

The `task log` for this long running tasks shows that the task status cycle between `waiting --> running` to `processing-->released` many many times. The processing resources aren't an issue. In some cases an error related to connection to the scheduler is printed (see below) so it seems that the worker (still up in the dashboard) looses the connection to the scheduler (temporarily or forever if the process stall). 

Interestingly enough this issue is less frequent if I build a graph and use `client.get` instead of using the `future api with client.gather`

```python

distributed.utils - ERROR - Timed out during handshake while connecting to tcp://192.168.0.2:35688 after 10 s                                      [112/221]
Traceback (most recent call last):
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/comm/core.py", line 319, in connect
    handshake = await asyncio.wait_for(comm.read(), time_left())
  File "/home/simone/mini/envs/test_d/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/utils.py", line 655, in log_errors
    yield
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/scheduler.py", line 3515, in retire_workers
    await self.replicate(
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/scheduler.py", line 3281, in replicate
    results = await asyncio.gather(
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/utils_comm.py", line 385, in retry_operation
    return await retry(
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/utils_comm.py", line 370, in retry
    return await coro()
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/core.py", line 880, in send_recv_from_rpc
    comm = await self.pool.connect(self.addr)
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/core.py", line 1031, in connect
    comm = await connect(
  File "/home/simone/mini/envs/test_d/lib/python3.8/site-packages/distributed/comm/core.py", line 324, in connect
    raise IOError(
OSError: Timed out during handshake while connecting to tcp://192.168.0.2:35688 after 10 s
```

It is difficult for me to provide a working minimal example because the issue happens only when the processing has been going on for quite a while (hrs) and not always after the same amount of time.

Any hint to what to look for in order to run a proper debugging will be helpful

Thanks.

**Anything else we need to know?**:

**Environment**:

- Dask version: 
dask                      2.14.0                     py_0    conda-forge
dask-core                 2.14.0                     py_0    conda-forge
dask-jobqueue             0.7.2                    pypi_0    pypi

or

dask                      2021.3.1           pyhd8ed1ab_0    conda-forge
dask-core                 2021.3.1           pyhd8ed1ab_0    conda-forge
dask-jobqueue             0.7.2                    pypi_0    pypi

- Python version: 3.8.5
- Operating System: linux
- Install method (conda, pip, source): conda

**Comment by simone-codeluppi** [comment]
Hi
that make sense. When I was running the processing building a graph and use `client.get` the task processing was shorter.

Here some info:
-  pdf with the log for one worker processing a stalled task :[task_log](https://www.dropbox.com/s/ngx3739hpv5ugmw/dask-out_Task_%20processing-fov-369.pdf?dl=0)
- pdf with overview info of the worker: [overview worker](https://www.dropbox.com/s/gwtvebuhoilbhkq/Worker_%20tcp___192.168.0.3_36982.pdf?dl=0)

- collected dask logs (I think scheduler logs): [scheduler](https://www.dropbox.com/s/y76dphcy0yks5id/pipeline_run_210330_12_50_29.log?dl=0)

- condor log/output examples: [condor_logs](https://www.dropbox.com/sh/jd7cluhj5jup0jq/AABpkvQtrUP2IzYyFJAy5nA9a?dl=0)

Example of a worker log
```python
distributed.nanny - INFO -         Start Nanny at: 'tcp://192.168.0.5:36607'
distributed.worker - INFO -       Start worker at:    tcp://192.168.0.5:36471
distributed.worker - INFO -          Listening to:    tcp://192.168.0.5:36471
distributed.worker - INFO -          dashboard at:          192.168.0.5:33390
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   10.00 GB
distributed.worker - INFO -       Local Directory: /tmp/dask-worker-space/worker-nxfz0d3o
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO -         Registered to:   tcp://193.10.16.58:10537
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.dask_worker - INFO - Exiting on signal 15
distributed.nanny - INFO - Closing Nanny at 'tcp://192.168.0.5:36607'
distributed.worker - INFO - Stopping worker at tcp://192.168.0.5:36471
distributed.nanny - INFO - Worker closed
distributed.dask_worker - INFO - End worker
```

Additional worker log
``` python
distributed.nanny - INFO -         Start Nanny at: 'tcp://192.168.0.32:44862'
distributed.worker - INFO -       Start worker at:   tcp://192.168.0.32:43008
distributed.worker - INFO -          Listening to:   tcp://192.168.0.32:43008
distributed.worker - INFO -          dashboard at:         192.168.0.32:41969
distributed.worker - INFO - Waiting to connect to:   tcp://193.10.16.58:10537
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   10.00 GB
distributed.worker - INFO -       Local Directory: /tmp/dask-worker-space/worker-7j87d7re
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:   tcp://193.10.16.58:10537
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
```

**Comment by guillaumeeb** [comment]
Hi,

I've no hint on what is going on there, but just to say that I've seen a similar behaviour on several workloads: tasks stall at random after several hours of computation. This is really a paint point for several of our Dask users. Sadly, I don't have a reproducible example.

I havnen't checked if tasks duration do increase over time before stalling, and didn't look on the logs yet.

**Comment by carusyte** [comment]
I must say I'm one of those users affected and found no solution for weeks if not months. I'm running LocalCluster on a single workstation machine so network issue is unlikely. It appeared that the worker processes started to get starved for tasks and entire cluster gets slower and lower throughput after like 10K of tasks. Each task takes about 10s to complete, CPU intensive.

After running for about 45min, I started to get 
```
Futures cancelled for reason: scheduler-connection-lost. 
Message: Client lost the connection to the scheduler. Please check your connection and re-run your work. 
```

And then many `Connection timed out` error followed. And most notably the event loops in scheduler and nanny are getting unresponsive increasingly

```
distributed.core - INFO - Event loop was unresponsive in Nanny for 4.41s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 4.41s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 4.41s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 4.41s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 4.41s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Scheduler for 14.70s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 33.84s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 33.84s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 43.46s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 55.30s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 53.26s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 53.26s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 53.26s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 53.26s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 53.26s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Scheduler for 42.22s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 24.01s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 24.01s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
distributed.core - INFO - Event loop was unresponsive in Nanny for 14.37s.  This is often caused by long-running GIL-holding functions or moving large chunks of data. This can cause timeouts and instability.
```

Although the task / workload per se didn't seem to degrade in performance, the scheduler event loop from the py-spy flamegraph seems struggling in creating tasks and add_future. Tried replacing tornado with asyncio but no luck.

![Image](https://github.com/user-attachments/assets/b1050085-b995-470a-85c3-df8add1ce3e6)

**Comment by jacobtomlinson** [comment]
Without some kind of reproducer it's very hard to debug issues like this.

**Comment by carusyte** [comment]
> Without some kind of reproducer it's very hard to debug issues like this.

Makes sense. Eventually I think I've managed to narrow down the possible causes and made a reproducer for my case in this new issue: #8999. Hope that helps

**Comment by jacobtomlinson** [comment]
That's great thanks @carusyte! I'm going to close this issue in favour of #8999.
